core/storage/blog_statistics/gae_models.py

```diff
@@ -0,0 +1,842 @@
+                base_models.ID_LENGTH)
+        return cls.get_new_id('%s:%s:%s' % (
+            utils.get_time_in_millisecs(timestamp),
+            blog_post_id,
+            random_hash)
+        )
+
+    @classmethod
+    def create(
+            cls,
+            blog_post_id: str
+    ) -> str:
+        """Creates a new blog post read event entry."""
+        entity_id = cls.get_new_event_entity_id(blog_post_id)
+        event_entity = cls(
+            id=entity_id,
+            blog_post_id=blog_post_id
+        )
+        event_entity.update_timestamps()
+        event_entity.put()
+        return entity_id
+
+    @staticmethod
+    def get_model_association_to_user() -> (
+        base_models.MODEL_ASSOCIATION_TO_USER
+    ):
+        """Model does not contain neccessary information for user for
+        takeout.
+        """
+        return base_models.MODEL_ASSOCIATION_TO_USER.NOT_CORRESPONDING_TO_USER
+
+    @classmethod
+    def get_export_policy(cls) -> Dict[str, base_models.EXPORT_POLICY]:
+        """Model doesn't contain any data directly corresponding to a user that
+        should be exported.
+        """
+        return dict(super(cls, cls).get_export_policy(), **{
+            'blog_post_id': base_models.EXPORT_POLICY.NOT_APPLICABLE
+        })
+
+
+class BlogPostExitedEventLogEntryModel(base_models.BaseModel):
+    """An event triggered when a blog post is read by any user, that is,
+    if the user stays on the blog post longer than 50% of the time calculated
+    using the number of words in the blog post,the blog post will be marked as
+    read.
+    The model will be keyed to unique ID which will be of the form -
+    [timestamp]:[blog_post_id]:[random_hash].
+    """
+
+    # ID of blog post being exited.
+    blog_post_id = datastore_services.StringProperty(
+        indexed=True, required=True)
+    # Time user stayed on the blog post.
+    time_user_stayed_on_blog_post = (
+        datastore_services.FloatProperty(indexed=True, required=True))
+
+    @staticmethod
+    def get_deletion_policy() -> base_models.DELETION_POLICY:
+        """Model doesn't contain any data directly corresponding to a user."""
+        return base_models.DELETION_POLICY.NOT_APPLICABLE
+
+    @classmethod
+    def get_new_event_entity_id(cls, blog_post_id: str) -> str:
+        """Generates a unique id for the event model of the form
+        '[timestamp]:[blog_post_id]:[random_hash]'.
+        """
+        timestamp = datetime.datetime.utcnow()
+
+        # To avoid collision between events occuring at the same date and time,
+        # a random hash is appended to the event model id.
+        for _ in range(base_models.MAX_RETRIES):
+            random_hash = utils.convert_to_hash(
+                str(utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+        return cls.get_new_id('%s:%s:%s' % (
+            utils.get_time_in_millisecs(timestamp),
```

Comment:
Same comment about timestamps as above. Also applicable everywhere else the code uses `utcnow`

Reply:
Done